### PR TITLE
Implemented sys_game_watchdog syscalls

### PR DESCRIPTION
**Pull request by brian218**
I implemented three sys_game_watchdog syscalls (sys_game_watchdog_start, sys_game_watchdog_stop, and sys_game_watchdog_clear) as a few games seem to rely on the watchdog to reboot the game itself.

### CODE REVIEW

**Comment by brian218** [comment]
Needs more suggestions

File: rpcs3/Emu/Cell/lv2/sys_game.cpp [hunk 1]
```diff
@@ -40,3 +49,50 @@ u64 _sys_game_get_system_sw_version()
...
+	{
+		while (!watchdog_stopped)
+		{
+			if (get_timestamp() - watchdog_last_clear > time * 1000 * 1000)
+			{
+				watchdog_stopped = true;
+				if (!Emu.IsStopped())
+				{
+					sys_game.warning("Watchdog timeout! Restarting the game...");
+					std::thread([](){Emu.Restart();}).detach();
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_game.cpp:67]: if this is already in a thread, why does it need another?

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_game.cpp]: Aha, right.

File: rpcs3/Emu/Cell/lv2/sys_game.cpp [hunk 2]
```diff
@@ -40,3 +51,50 @@ u64 _sys_game_get_system_sw_version()
 {
 	return stof(utils::get_firmware_version()) * 10000;
 }
+
+error_code _sys_game_watchdog_start(u32 time)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_game.cpp:55]: what if this is called multiple times?
you'll spawn thousands of threads...

File: rpcs3/Emu/Cell/lv2/sys_game.cpp [hunk 3]
```diff
@@ -2,12 +2,23 @@
...
 #include "Emu/system_config.h"
 
+#include <thread>
+
 #include "sys_game.h"
 
 LOG_CHANNEL(sys_game);
 
+bool watchdog_stopped;
+u64 watchdog_last_clear;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_game.cpp:16]: these should probably be atomics or something

File: rpcs3/Emu/Cell/lv2/sys_game.cpp (hunk 2 above)
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_game.cpp:55]: > what if this is called multiple times? you'll spawn thousands of threads...

In the log, I can see that the game just called this function once when booting.
It should be the game's responsibility to make sure this function is used properly.

File: rpcs3/Emu/Cell/lv2/sys_game.cpp (hunk 3 above)
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_game.cpp:16]: > these should probably be atomics or something

I agree. Could you provide some example implementation of atomics?

File: rpcs3/Emu/Cell/lv2/sys_game.cpp (hunk 2 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/lv2/sys_game.cpp:55]: so you want the emulator to restart 1000 times at once if the game decides to be weird?

I'd just add a bool that checks if a watchdog exists.
Or use an fxo object that handles this.

File: rpcs3/Emu/Cell/lv2/sys_game.cpp (hunk 3 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/lv2/sys_game.cpp:16]: you can just use:
```
atomic_t<bool> watchdog_stopped;
atomic_t<u64> watchdog_last_clear;
```

File: rpcs3/Emu/Cell/lv2/sys_game.cpp [hunk 4]
```diff
@@ -40,3 +51,55 @@ u64 _sys_game_get_system_sw_version()
...
+	if (!watchdog_stopped)
+	{
+		return CELL_EABORT;
+	}
+
+	auto watchdog = [=]()
+	{
+		while (!watchdog_stopped)
+		{
+			if (get_timestamp() - watchdog_last_clear > time * 1000 * 1000)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_game.cpp:68]: I think "time" should be called "timeout". Preferable with the proper unit as suffix in the name.

File: rpcs3/Emu/Cell/lv2/sys_game.cpp (hunk 3 above)
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_game.cpp]: Thanks!

File: rpcs3/Emu/Cell/lv2/sys_game.cpp (hunk 4 above)
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_game.cpp]: Updated.

[the remark this answers is not in the document]
Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_game.cpp]: Added.

File: rpcs3/Emu/Cell/lv2/sys_game.cpp [hunk 5]
```diff
@@ -35,8 +105,3 @@ error_code _sys_game_board_storage_read(vm::ptr<u8> buffer1, vm::ptr<u8> buffer2
 
 	return CELL_OK;
 }
-
-u64 _sys_game_get_system_sw_version()
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/sys_game.cpp:39]: Why is this moved?

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_game.cpp:39]: I just re-arranged the order according to the order of sys_game syscalls in lv2.cpp.

**Comment by brian218** [comment]
> Uh, there are thread safety issues and bugs with emulation stopping at this state. I'm also not sure why Emu.Restart() was choosed over Emu.Stop().

The watchdog is supposed to restart the game. Also, I've observed that some arcade games rely on the watchdog to make the game itself restarted, by stopping calling sys_game_watchdog_clear().
As for the thread safety, I adopted `Emu.CallFromMainThread([](){Emu.Restart();});` so the Emu.Restart() will be called from the GUI thread, as if the player clicked "Restart" in the RPCS3 menu -> Emulation.

**Comment by elad335** [comment]
As for the emulation stopping bug, I was referring to the fact that the watchdog thread operation leaks even after GUI emulation stopping request ie veing issued, potentially restarting other applications in this time frame.

**Comment by brian218** [comment]
> As for the emulation stopping bug, I was referring to the fact that the watchdog thread operation leaks even after GUI emulation stopping request ie veing issued, potentially restarting other applications in this time frame.

How about changing `std::this_thread::sleep_for(1s);` to a smaller value? or adding _sys_game_watchdog_stop() to the game shutdown process of RPCS3?

**Comment by elad335** [comment]
Neither, it should be manged by a named_thread held by FXO and joined on emulation stopping implicitly by its destructor.

**Comment by elad335** [comment]
```Also, the watchdog won't issue the restart request when Emu.IsStopped() == true.```
This is also a race condition.

**Comment by brian218** [comment]
> Neither, it should be manged by a named_thread held by FXO and joined on emulation stopping implicitly by its destructor.

Hmm. Maybe I can make another PR to optimize it later. Would you mind providing some examples of your suggestion?

**Comment by brian218** [comment]
> > Neither, it should be manged by a named_thread held by FXO and joined on emulation stopping implicitly by its destructor.
> 
> Hmm. Maybe I can make another PR to optimize it later. Would you mind providing some examples of your suggestion?

Any examples would be appreciated.

**Comment by brian218** [comment]
How about the following snippets?
`struct watchdog
{
	atomic_t<bool> stopped = true;
	atomic_t<u64> last_clear;
	std::thread monitor;

	watchdog(const watchdog&) = delete;
	watchdog& operator=(const watchdog&) = delete;
};`
`
	g_fxo->get<watchdog>().stopped = false;
	g_fxo->get<watchdog>().last_clear = get_timestamp();
	g_fxo->get<watchdog>().monitor = std::thread(watchdog_thread);
`
